add Python 3.12 support to ansible-test

s-hertel opened the pull request:
##### SUMMARY

add Python 3.12 as a supported ansible-test controller Python version 

update test containers

add Python 3.12 to the text matrix

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
Prime ansible-test for installing Python 3.12 in the default test containers

s-hertel committed:
update docker containers versions to use newer ansible-test ref in the pre-built venvs

s-hertel committed:
Allow invoking ansible-test with Python 3.12

s-hertel committed:
Add python3.12 to the INTERPRETER_PYTHON_FALLBACK

s-hertel committed:
changelog

s-hertel committed:
add Python 3.12 as a non-default Python version for the test containers

mattclay commented on test/sanity/code-smell/docs-build.json:
```diff
@@ -1,5 +1,6 @@
     "no_targets": true,
-    "output": "path-line-column-message"
+    "output": "path-line-column-message",
+    "maximum_python_version": "3.11"
```
This change is no longer needed, now that the `docs-build` sanity test has been removed.

s-hertel committed:
* Update unit test for using the Python 2 collection loader path with
Python 3.

Skip the existing test on Python 3.12, since find_module is removed.

Suppress the pre-existing deprecation warnings using the Python 2
codepath with Python 3.

Add a test for Python >= 3.12, which doesn't call find_module.

* Ignore sanity test errors on systems without libselinux present.

s-hertel committed:
Update mypy ignores for Python 3.12

mattclay commented:
@s-hertel Python 3.12 should be added to the AZP config where we already test Python 3.11:

https://github.com/ansible/ansible/blob/08307de366338c6bdae6aab98d8df2e0630dff94/.azure-pipelines/azure-pipelines.yml#L183

https://github.com/ansible/ansible/blob/08307de366338c6bdae6aab98d8df2e0630dff94/.azure-pipelines/azure-pipelines.yml#L193

https://github.com/ansible/ansible/blob/08307de366338c6bdae6aab98d8df2e0630dff94/.azure-pipelines/azure-pipelines.yml#L63

s-hertel committed:
Add Python 3.12 to CI matrix for unit tests, generic tests, and galaxy

ci_complete

mattclay commented:
The collection loader unit test needs a minor update:

https://github.com/ansible/ansible/blob/59a791ee3be41d00f65fe47c20dd93489dd7222b/test/units/utils/collection_loader/test_collection_loader.py#L40

The assert should suppress the `DeprecationWarning` it triggers on Python 3.11 and earlier. On Python 3.12 that assert needs to be skipped since `find_module` has been removed from that Python version.

mattclay commented on test/units/utils/collection_loader/test_collection_loader.py:
```diff
@@ -37,7 +37,9 @@ def test_find_module_py3():
     assert path_hook_finder.find_spec('missing') is None
-    assert path_hook_finder.find_module('missing') is None
+    if sys.version_info < (3, 12):
+        # find_module has been removed in Python 3.12
```
Since this generates a deprecation warning on Python 3.11, we should suppress that here.

s-hertel commented on test/support/windows-integration/plugins/action/win_reboot.py:
```diff
@@ -64,7 +64,7 @@ def perform_reboot(self, task_vars, distribution):
         result = {}
         reboot_result = self._low_level_execute_command(reboot_command, sudoable=self.DEFAULT_SUDOABLE)
-        result['start'] = datetime.utcnow()
+        resutl['start'] = datetime.now(timezone.utc)
```
```suggestion
        result['start'] = datetime.now(timezone.utc)
```

webknjaz commented on test/units/utils/collection_loader/test_collection_loader.py:
```diff
@@ -37,7 +38,11 @@ def test_find_module_py3():
-    assert path_hook_finder.find_module('missing') is None
+    if sys.version_info < (3, 12):
+        # removed in Python 3.12
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", DeprecationWarning)
```
@s-hertel could you replace this with a decorator following pytest's instructions and make it more accurate?

It'd looks like this:
```python
@pytest.mark.filterwarnings(
    'ignore:'
    r'the load_module\(\) method is deprecated and slated for removal '
    r'in Python 3\.12; use exec_module\(\) instead'
    ':DeprecationWarning',
)
```

Ref: https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#pytest-mark-filterwarnings

mattclay replied:
If we're filtering warnings for the entire function using a decorator, we should probably split out the assert into its own test so the scope of the warning suppression is limited to the assert it's intended for.

s-hertel replied:
good idea, split into two tests

webknjaz commented on test/units/utils/collection_loader/test_collection_loader.py:
```diff
@@ -37,7 +38,11 @@ def test_find_module_py3():
     # see https://github.com/pypa/setuptools/pull/2918
     assert path_hook_finder.find_spec('missing') is None
-    assert path_hook_finder.find_module('missing') is None
+    if sys.version_info < (3, 12):
```
Let's invert this to dedent stuff below:
```suggestion
    if sys.version_info >= (3, 12):
        return
```

mattclay replied:
If this assert ends up in a separate test (see my previous comment), the the version check can also be done with a decorator.

webknjaz commented on changelogs/fragments/ansible-test-python-3.12.yml:
```diff
@@ -0,0 +1,4 @@
+minor_changes:
+  - ansible-test - Add support for testing with Python 3.12.
+  - ansible-test - Update pip and setuptools versions.
```
Can we specify the versions in the change log? I think, this information can be useful.

s-hertel replied:
added

mattclay commented on test/sanity/ignore.txt:
```diff
@@ -6,10 +6,13 @@ lib/ansible/executor/powershell/async_wrapper.ps1 pslint:PSCustomUseLiteralPath
 lib/ansible/executor/powershell/exec_wrapper.ps1 pslint:PSCustomUseLiteralPath
 lib/ansible/galaxy/collection/__init__.py mypy-3.10:attr-defined  # inline ignore has no effect
 lib/ansible/galaxy/collection/__init__.py mypy-3.11:attr-defined  # inline ignore has no effect
+lib/ansible/galaxy/collection/__init__.py mypy-3.12:attr-defined  # inline ignore has no effect
```
This file should also include a skip entry for Python 3.12 like the one here:

https://github.com/ansible/ansible/blob/e964078a83af240b103e34eafca9162deff0e16f/test/sanity/ignore.txt#L51

That will avoid sanity test errors on systems without libselinux present.

s-hertel replied:
There isn't one for 3.11 or 3.10, so I didn't add one for 3.12. Should I add those as well? https://github.com/ansible/ansible/blob/e964078a83af240b103e34eafca9162deff0e16f/test/sanity/ignore.txt#L51-L55.

mattclay replied:
I already added the missing 3.10 and 3.11 entries in https://github.com/ansible/ansible/pull/81305

mattclay commented on test/units/utils/collection_loader/test_collection_loader.py:
```diff
@@ -40,6 +48,16 @@ def test_find_module_py3():
     assert path_hook_finder.find_module('missing') is None
 
 
+@pytest.mark.skipif(not PY3 or sys.version_info < (3, 12), reason='Testing Python 2 codepath (find_module) on Python >= 3.12')
```
This could be simplified:

```suggestion
@pytest.mark.skipif(sys.version_info < (3, 12), reason='Testing Python 2 codepath (find_module) on Python >= 3.12')
```

s-hertel commented:
/azp run

s-hertel commented:
@mattclay I think this is ready to merge. I separated the datetime stuff into https://github.com/ansible/ansible/pull/81323 since it could probably be backported.
